### PR TITLE
Implement the `updateData` and `setData` methods alongside `loadData`.

### PR DESCRIPTION
**Pull request by jansiegel**
### Context
This PR implements two new methods:
- `updateData` - Works like `loadData`, but without resetting the cell meta and index mapper information.
- `setData` - basically an alias to `loadData`

The `loadData` method should work unchanged to keep the backward compatibility.

More information in [this comment](https://github.com/handsontable/handsontable/pull/8972#issuecomment-979184070).

#### Other changes:
- The `updateSettings` method uses `setData` only when initializing Handsontable, in all other cases it utilizes `updateData`. This means, that calling `updateSettings` with a `data` option should no longer reset metadata and index mapper information.
- For backward compatibility, `updateSettings` will still trigger the `beforeLoadData` and `afterLoadData` hooks, but they should probably be marked as deprecated (now or in the future?), as they're duplicating the functionality of the `setData`/`updateData` hooks.

#### Left to do:
- [x] Add backward compatibility aliases for the `setData`/`loadData` hooks in the plugins (currently they'll only react to `setData` hooks, but calling `loadData` externally will not work for most of them).
- [x] Add the new functions to the TS definition files
- [x] Add more test cases.
- [x] Test this PR against the wrappers. (no major changes - replacing `updateSettings` calls with `updateData` _might_ be a good idea)

### How has this been tested?
Added new test cases and modified old ones, that did not match the new logic.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):
1. #7263 

### Affected project(s):
- [x] `handsontable`
- [x] `@handsontable/angular`
- [x] `@handsontable/react`
- [x] `@handsontable/vue`

### Checklist:

- [x] I have reviewed the guidelines about [Contributing to Handsontable](https://github.com/handsontable/handsontable/blob/master/CONTRIBUTING.md) and I confirm that my code follows the code style of this project.
- [x] I have signed the [Contributor License Agreement](https://docs.google.com/forms/d/e/1FAIpQLScpMq4swMelvw3-onxC8Jl29m0fVp5hpf7d1yQVklqVjGjWGA/viewform?c=0&w=1)
- [ ] My change requires a change to the documentation.

### CODE REVIEW

**Comment by jansiegel** [comment]
## Changes from this PR:

### Cell meta and Index Mappers

#### Before
1. Cell meta was being reset after calling the `updateSettings` method with a `data` option enabled.
https://jsfiddle.net/js_ziggle/59aLd6pq/
2. Index mappers were being reset after calling the `updateSettings` method with a `data` option enabled.
https://jsfiddle.net/js_ziggle/z13L58op/

#### After
1. Cell meta ins NOT being reset after calling the `updateSettings` method with a `data` option enabled.
https://jsfiddle.net/js_ziggle/jcdynrxg/
2. Index mappers are NOT being reset after calling the `updateSettings` method with a `data` option enabled. (this doesn't work when moving rows/columns with the `manualRowMove`/`manualColumnMove` plugin - caused by the issue with `updatePlugin` handling)
https://jsfiddle.net/js_ziggle/xncby1qt/

**_tl; dr;_**
- `loadData` becomes an alias for `setData`.
- `updateData` doesn't reset cell meta cache nor reinitialize the index mappers.

### `updateSettings`

#### Before
1. The `updateSettings` method, when provided with the `data` config option, called the `loadData` method, which reinitialized both the cell meta cache and the index mappers.
2. The `loadData` method (and thus `updateSettings` as well) triggered the following hooks:
    1. `beforeLoadData`
    2. `afterLoadData`

#### After
1. The `updateSettings` method, when provided with the `data` config option:
    1. Calls the `setData` method if it's the first call of the lifecycle (triggered by HOT initialization). `setData` is an equivalent of the currently-existing `loadData` method (it initializes the cell meta cache and index mappers).
    2. Calls the `updateMeta` method otherwise. `updateMeta` doesn't change the cell meta cache and modifies the index mappers only to trim them to the new desired sizes.
2. The `loadData` method (NOT used by `updateSettings`) triggers the following hooks:
    1. `beforeLoadData`
    2. `afterLoadData`
3. The `setData` method called directly triggers the following hooks:
    1. `beforeSetData`
    2. `afterSetData`
3. The `setData` method called from `updateSettings` triggers the following hooks:
    1. `beforeSetData`
    2. `beforeLoadData`
    3. `afterSetData`
    4. `afterLoadData`
3. The `updateData` method called directly triggers the following hooks:
    1. `beforeUpdateData`
    2. `afterUpdateData`
3. The `updateData` method called from `updateSettings` triggers the following hooks:
    1. `beforeUpdateData`
    2. `beforeLoadData`
    3. `afterUpdateData`
    4. `afterLoadData`

**_tl; dr;_**
- `loadData` is no longer used by `updateSettings`, which now triggers `setData` at init and `updateData` in all other situations.
- Each of the data-handling methods (`loadData`, `setData` and `updateMeta`) call their own specific `before-` and `after-` hooks, with the exception of `setData` and `updateMeta` which trigger `beforeLoadData`+`afterLoadData` as well, for maintaining backward compatibility.

#### Additional changes
- Renamed all the `before/afterLoadData` hook listeners in the plugins to use `setData` + added additional `before/afterLoadData` hook listeners, acting as an alias for the old ones, to not break the backward compatibility.

File: handsontable/test/e2e/Core_loadData.spec.js [hunk 1]
```diff
@@ -263,28 +263,34 @@ describe('Core_loadData', () => {
...
       rowHeights: [23, 23, 23, 23],
       cells: cellsSpy,
     });
 
-    expect(cellsSpy.calls.count()).toBe(12);
+    // Default `cells` calls (table initializes with default values)
+    const afterInitCellsSpy = cellsSpy.calls.count();
+
+    loadData(arrayOfNestedObjects());
+    expect(cellsSpy.calls.count() - afterInitCellsSpy).toBe(12);
```
Comment (budnix) [inline handsontable/test/e2e/Core_loadData.spec.js:275]:
```suggestion
    loadData(arrayOfNestedObjects());
    
    expect(cellsSpy.calls.count() - afterInitCellsSpy).toBe(12);
```

File: handsontable/test/e2e/Core_loadData.spec.js [hunk 2]
```diff
@@ -263,28 +263,34 @@ describe('Core_loadData', () => {
...
       rowHeights: [90, 90, 90, 90],
       cells: cellsSpy
     });
 
-    expect(cellsSpy.calls.count()).toBe(12);
+    // Default `cells` calls (table initializes with default values)
+    const afterInitCellsSpy = cellsSpy.calls.count();
+
+    loadData(arrayOfNestedObjects());
+    expect(cellsSpy.calls.count() - afterInitCellsSpy).toBe(12);
```
Comment (budnix) [inline handsontable/test/e2e/Core_loadData.spec.js:293]:
```suggestion
    loadData(arrayOfNestedObjects());
    
    expect(cellsSpy.calls.count() - afterInitCellsSpy).toBe(12);
```

File: handsontable/test/types/settings.types.ts [hunk 3]
```diff
@@ -448,6 +448,7 @@ const allSettings: Required<Handsontable.GridSettings> = {
   afterSelectionEnd: (r, c, r2, c2, selectionLayerLevel) => {},
   afterSelectionEndByProp: (r, p, r2, p2, selectionLayerLevel) => {},
```
Comment (budnix) [inline handsontable/test/types/settings.types.ts:449]: Thanks for updating the TS tests. However, can I ask you to add new methods tests to the `handsontable/test/types/methods.types.ts` file? There is no cover for `setData` and `updateData` methods.

File: handsontable/src/translations/indexMapper.js [hunk 4]
```diff
@@ -437,6 +437,26 @@ export class IndexMapper {
     this.runLocalHooks('init');
   }
 
+  /**
+   * Trim/extend the mappers to fit the desired length.
+   *
+   * @param {number} length New mapper length.
+   */
+  fitToLength(length) {
```
Comment (budnix) [inline handsontable/src/translations/indexMapper.js:445]: I've seen that you cover the change with [the e2e test](https://github.com/handsontable/handsontable/pull/8972/files#diff-b44f35a1b74ccc2a1eee272e7e75e690555997a55c3685375eefbc4128489e63R607) but there is no unit test for that specific module. Can you add one? Probably there `handsontable/src/translations/__tests__/indexMapper.unit.js`.

File: handsontable/src/core.js [hunk 5]
```diff
@@ -2101,27 +2117,100 @@ export default function Core(rootElement, userSettings, rootInstanceSymbol = fal
...
-    metaManager.clearCellsCache();
-    instance.initIndexMappers();
+    if (firstRun || resetState) {
+      metaManager.clearCellsCache();
+      instance.initIndexMappers();
+
+    } else {
+      this.columnIndexMapper.fitToLength(this.getInitialColumnCount());
+
+      this.rowIndexMapper.fitToLength(this.countSourceRows());
```
Comment (budnix) [inline handsontable/src/core.js:2127]:
```suggestion
      this.columnIndexMapper.fitToLength(this.getInitialColumnCount());
      this.rowIndexMapper.fitToLength(this.countSourceRows());
```

**Review by budnix — CHANGES_REQUESTED** [review]

File: handsontable/src/core.js [hunk 6]
```diff
@@ -2019,19 +2019,29 @@ export default function Core(rootElement, userSettings, rootInstanceSymbol = fal
...
    * @fires Hooks#beforeLoadData
+   * @fires Hooks#beforeSetData
+   * @fires Hooks#beforeUpdateData
    * @fires Hooks#afterLoadData
+   * @fires Hooks#afterSetData
+   * @fires Hooks#afterUpdateData
    * @fires Hooks#afterChange
    */
-  this.loadData = function(data, source) {
+  this.replaceData = function(data, source, internalSource, resetState = true) {
```
[the remark this answers is not in the document]
Comment (jansiegel) [reply handsontable/src/core.js:2040]: Good idea 👍

File: handsontable/src/core.js [hunk 7]
```diff
@@ -2018,110 +2019,114 @@ export default function Core(rootElement, userSettings, rootInstanceSymbol = fal
...
+   * @param {Array} data Array of arrays or array of objects containing data.
+   * @param {string} [source] Source of the loadData call.
+   * @fires Hooks#beforeLoadData
+   * @fires Hooks#beforeSetData
+   * @fires Hooks#afterLoadData
+   * @fires Hooks#afterSetData
+   * @fires Hooks#afterChange
+   */
+  this.loadData = function(data, source) {
+    this.setData(data, source, 'loadData');
```
Comment (budnix) [inline handsontable/src/core.js:2120]: I think that the 3rd argument is still problematic here. As the [comment](https://github.com/handsontable/handsontable/pull/8972#discussion_r757325482) describes I can invoke `hot.setData([[1, 2, 3]], 'edit', 'columnCollapse')` in result the `beforeColumnCollapse` hook is called 🙀

I would be in favor of copy&paste the whole `replaceData` call from `setData` and paste into `loadData` with `internalSource: 'loadData'` then the `/* eslint-enable jsdoc/require-param */` rule could be removed.

File: handsontable/src/dataMap/replaceData.js [hunk 8]
```diff
@@ -0,0 +1,142 @@
+import { toUpperCaseFirst } from '../helpers/string';
+import { isFunction } from '../helpers/function';
+import { DataMap } from './index';
```
Comment (budnix) [inline handsontable/src/dataMap/replaceData.js:3]:
```suggestion
import DataMap from './dataMap';
```

File: handsontable/test/e2e/Core_loadData.spec.js [hunk 9]
```diff
@@ -139,14 +138,14 @@ describe('Core_loadData', () => {
         { data: 'full.street' },
       ]
     });
+    loadData(arrayOfNestedObjects());
```
Comment (budnix) [inline handsontable/test/e2e/Core_loadData.spec.js:141]:
```suggestion
    loadData(arrayOfNestedObjects());
    
```

File: handsontable/test/e2e/Core_loadData.spec.js [hunk 10]
```diff
@@ -166,16 +165,17 @@ describe('Core_loadData', () => {
         return colMeta;
       }
     });
+    loadData(arrayOfNestedObjects());
```
Comment (budnix) [inline handsontable/test/e2e/Core_loadData.spec.js:168]:
```suggestion
    loadData(arrayOfNestedObjects());
    
```

File: handsontable/test/e2e/Core_loadData.spec.js [hunk 11]
```diff
@@ -166,16 +165,17 @@ describe('Core_loadData', () => {
...
     expect(getDataAtCell(1, 3)).toEqual('Street II');
     expect(getDataAtRowProp(2, 'full.street')).toEqual('Street III');
   });
 
   it('should figure out default column names for array of nested objects', () => {
     handsontable({
-      data: arrayOfNestedObjects(),
       colHeaders: true
     });
+    loadData(arrayOfNestedObjects());
```
Comment (budnix) [inline handsontable/test/e2e/Core_loadData.spec.js:178]:
```suggestion
    loadData(arrayOfNestedObjects());
    
```

File: handsontable/test/e2e/Core_loadData.spec.js [hunk 12]
```diff
@@ -227,18 +227,18 @@ describe('Core_loadData', () => {
   it('should create new rows for array of arrays (and respect minRows)', () => {
     handsontable({
       minRows: 20, // minRows should be respected
-      data: arrayOfArrays()
     });
 
+    loadData(arrayOfArrays());
```
Comment (budnix) [inline handsontable/test/e2e/Core_loadData.spec.js:232]:
```suggestion
    loadData(arrayOfArrays());
    
```

File: handsontable/test/e2e/Core_loadData.spec.js [hunk 13]
```diff
@@ -227,18 +227,18 @@ describe('Core_loadData', () => {
...
     expect(countRows()).toEqual(20); // TODO why this must be checked after render?
   });
 
   it('should create new rows for array of nested objects (and respect minRows)', () => {
     handsontable({
       minRows: 20, // minRows should be respected
-      data: arrayOfNestedObjects()
     });
 
+    loadData(arrayOfNestedObjects());
```
Comment (budnix) [inline handsontable/test/e2e/Core_loadData.spec.js:241]:
```suggestion
    loadData(arrayOfNestedObjects());
    
```

File: handsontable/test/e2e/Core_loadData.spec.js [hunk 14]
```diff
@@ -252,9 +252,9 @@ describe('Core_loadData', () => {
   it('should create as many rows as needed by array of objects', () => {
     handsontable({
       minRows: 6,
-      data: arrayOfObjects()
     });
 
+    loadData(arrayOfObjects());
```
Comment (budnix) [inline handsontable/test/e2e/Core_loadData.spec.js:257]:
```suggestion
    loadData(arrayOfObjects());
    
```

File: handsontable/test/e2e/Core_loadData.spec.js [hunk 15]
```diff
@@ -382,21 +393,20 @@ describe('Core_loadData', () => {
...
 
     handsontable({
-      data: data1,
       columns: [
         { data: 1 },
         { data: 3 }
       ]
     });
 
+    loadData(data1);
```
Comment (budnix) [inline handsontable/test/e2e/Core_loadData.spec.js:402]:
```suggestion
    loadData(data1);
    
```

File: handsontable/test/e2e/Core_loadData.spec.js [hunk 16]
```diff
@@ -410,23 +420,10 @@ describe('Core_loadData', () => {
       }
     });
 
+    loadData(data1);
```
Comment (budnix) [inline handsontable/test/e2e/Core_loadData.spec.js:423]:
```suggestion
    loadData(data1);
    
```

File: handsontable/types/translations/indexMapper.d.ts [hunk 17]
```diff
@@ -22,6 +22,7 @@ export class IndexMapper {
   registerMap(uniqueName: string, indexMap: IndexMap): IndexMap;
   unregisterMap(name: string): void;
   unregisterAll(): void;
+  fitToLength(length?: number): void;
```
Comment (budnix) [inline handsontable/types/translations/indexMapper.d.ts:25]:
```suggestion
  fitToLength(length: number): void;
```

**Review by budnix — CHANGES_REQUESTED** [review]

File: handsontable/src/core.js (hunk 7 above)
Comment (jansiegel) [reply handsontable/src/core.js:2120]: Not ideal, but you're probably right, that it'd be the safest approach.

**Review by budnix — APPROVED** [review]
Thanks 🥇

**Review by budnix — APPROVED** [review]

**Pull request merged**